src/prefect/server/api/concurrency_limits.py

```diff
@@ -188,79 +385,163 @@ async def increment_concurrency_limits_v1(
     ),
     db: PrefectDBInterface = Depends(provide_database_interface),
 ) -> List[MinimalConcurrencyLimitResponse]:
-    applied_limits = {}
-
+    # Check for V2 limits
+    v2_names = [f"tag:{tag}" for tag in names]
     async with db.session_context(begin_transaction=True) as session:
-        try:
-            applied_limits = {}
-            filtered_limits = (
-                await concurrency_limits.filter_concurrency_limits_for_orchestration(
-                    session, tags=names
-                )
+        v2_limits = await cl_v2_models.bulk_read_concurrency_limits(
+            session=session, names=v2_names
+        )
+        v2_by_name = {limit.name: limit for limit in v2_limits}
+
+        # Check for V1 limits
+        v1_limits = (
+            await concurrency_limits.filter_concurrency_limits_for_orchestration(
+                session, tags=names
+            )
+        )
+        v1_by_tag = {limit.tag: limit for limit in v1_limits}
+
+        # Process each requested tag
+        applied_v2_ids = []
+        applied_v1_tags = []
+        results = []
+
+        for tag in names:
+            v2_limit = v2_by_name.get(f"tag:{tag}")
+            v1_limit = v1_by_tag.get(tag)
+
+            # Prefer V2 if it exists
+            TAG_CONCURRENCY_SLOT_WAIT_SECONDS = (
+                get_current_settings().server.tasks.tag_concurrency_slot_wait_seconds
             )
-            run_limits = {limit.tag: limit for limit in filtered_limits}
-            for tag, cl in run_limits.items():
-                limit = cl.concurrency_limit
-                if limit == 0:
-                    # limits of 0 will deadlock, and the transition needs to abort
-                    for stale_tag in applied_limits.keys():
-                        stale_limit = run_limits.get(stale_tag, None)
-                        active_slots = set(stale_limit.active_slots)
+            if v2_limit:
+                # Check for zero limit
+                if v2_limit.limit == 0:
```

Comment:
I think we could reduce the complexity by checking if there are any zero limits before incrementing anything.